In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
#!pip install pycaret
from pycaret.classification import *
from pycaret.regression import *

# Importing the data - only use *.MLReady.csv files!!!

In [2]:
df=pd.read_csv("data/FannieAcq2000-19FMac30samp100QuickenMlReady14cut.csv")
df.head()

Unnamed: 0  origChannel  origIntRate  origUPB  origLTV  numBorrowers  \
0           0            1        4.750    84000       53             1   
1           1            1        4.875   137000       69             1   
2           2            1        4.750   459000       77             2   
3           3            1        4.250   219000       77             1   
4           4            1        5.250    92000       80             1   

   origDebtIncRatio  loanPurp  zipCode  pMIperct  ...  fmacRate  fredRate  \
0                44         1      724       0.0  ...      4.32      2.67   
1                47         1      802       0.0  ...      4.32      2.67   
2                42         2      109       0.0  ...      4.32      2.67   
3                22         2      956       0.0  ...      4.32      2.67   
4                44         1      760       0.0  ...      4.32      2.67   

   rateDiffPct  rank      deal  bankNumber stateNumber  origYear  origMonth  \
0     0.099537     4   Premium           0           2      2014          2   
1     0.128472     4   Premium           0           5      2014          2   
2     0.099537     4   Premium           0          34      2014          2   
3    -0.016204     3  Discount           0           4      2014          2   
4     0.215278     4   Premium           0          43      2014          2   

   origWeek  
0         5  
1         5  
2         5  
3         5  
4         5  

[5 rows x 22 columns]

In [3]:
df.drop(['Unnamed: 0','bankNumber'],1,inplace=True)

In [4]:
df.head()

origChannel  origIntRate  origUPB  origLTV  numBorrowers  origDebtIncRatio  \
0            1        4.750    84000       53             1                44   
1            1        4.875   137000       69             1                47   
2            1        4.750   459000       77             2                42   
3            1        4.250   219000       77             1                22   
4            1        5.250    92000       80             1                44   

   loanPurp  zipCode  pMIperct  mortInsType  avgCreditScore  fmacRate  \
0         1      724       0.0          0.0             733      4.32   
1         1      802       0.0          0.0             799      4.32   
2         2      109       0.0          0.0             753      4.32   
3         2      956       0.0          0.0             766      4.32   
4         1      760       0.0          0.0             707      4.32   

   fredRate  rateDiffPct  rank      deal  stateNumber  origYear  origMonth  \
0      2.67     0.099537     4   Premium            2      2014          2   
1      2.67     0.128472     4   Premium            5      2014          2   
2      2.67     0.099537     4   Premium           34      2014          2   
3      2.67    -0.016204     3  Discount            4      2014          2   
4      2.67     0.215278     4   Premium           43      2014          2   

   origWeek  
0         5  
1         5  
2         5  
3         5  
4         5

In [5]:
#Checking datatypes of individual feature
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 293383 entries, 0 to 293382
Data columns (total 20 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   origChannel       293383 non-null  int64  
 1   origIntRate       293383 non-null  float64
 2   origUPB           293383 non-null  int64  
 3   origLTV           293383 non-null  int64  
 4   numBorrowers      293383 non-null  int64  
 5   origDebtIncRatio  293383 non-null  int64  
 6   loanPurp          293383 non-null  int64  
 7   zipCode           293383 non-null  int64  
 8   pMIperct          293383 non-null  float64
 9   mortInsType       293383 non-null  float64
 10  avgCreditScore    293383 non-null  int64  
 11  fmacRate          293383 non-null  float64
 12  fredRate          293383 non-null  float64
 13  rateDiffPct       293383 non-null  float64
 14  rank              293383 non-null  int64  
 15  deal              293383 non-null  object 
 16  stateNumber       29

In [6]:
#Dropping columns we don't want
# df.drop(['loanIdentifier','namedCol', 'firstPmtDate', 'bestCreditScore', 'worstCreditScore', 'rateDiff',
#            'borrCreditScore'],1,inplace=True)

In [7]:
#df.to_csv('data/FannieAcq2000-19FMac30-sample2pt5_MlReady.csv')

# Review values for dates and slice (if needed) the timeframe you want

In [8]:
df.origYear.unique()

array([2014, 2015, 2016, 2017, 2018, 2019], dtype=int64)

In [9]:
df['origMonth'].unique()

array([ 2,  1,  3,  6,  4,  5,  8,  9,  7, 11, 10, 12], dtype=int64)

## Example slice from Oct 2013 forward

In [10]:
# df = df[df['origYear'] >= 2013]


In [11]:
# df = df[~((df['origYear'] == 2013) & (df['origMonth'] < 10))]

In [12]:
# df.shape

In [13]:
# df.origMonth.unique()

In [14]:
# df.shape

In [15]:
# df.info()

# Coerce the column dtype (int in this case)

In [16]:
df['origUPB'] = df['origUPB'].astype(int)
df['origLTV'] = df['origLTV'].astype(int)
df['origDebtIncRatio'] = df['origDebtIncRatio'].astype(int)
df['avgCreditScore'] = df['avgCreditScore'].astype(int)
df['numBorrowers'] = df['numBorrowers'].astype(int)
df['rank'] = df['rank'].astype(int)

# Review metrics last time before run (to have record of starting values)

In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 293383 entries, 0 to 293382
Data columns (total 20 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   origChannel       293383 non-null  int64  
 1   origIntRate       293383 non-null  float64
 2   origUPB           293383 non-null  int32  
 3   origLTV           293383 non-null  int32  
 4   numBorrowers      293383 non-null  int32  
 5   origDebtIncRatio  293383 non-null  int32  
 6   loanPurp          293383 non-null  int64  
 7   zipCode           293383 non-null  int64  
 8   pMIperct          293383 non-null  float64
 9   mortInsType       293383 non-null  float64
 10  avgCreditScore    293383 non-null  int32  
 11  fmacRate          293383 non-null  float64
 12  fredRate          293383 non-null  float64
 13  rateDiffPct       293383 non-null  float64
 14  rank              293383 non-null  int32  
 15  deal              293383 non-null  object 
 16  stateNumber       29

In [18]:
df.head()

origChannel  origIntRate  origUPB  origLTV  numBorrowers  origDebtIncRatio  \
0            1        4.750    84000       53             1                44   
1            1        4.875   137000       69             1                47   
2            1        4.750   459000       77             2                42   
3            1        4.250   219000       77             1                22   
4            1        5.250    92000       80             1                44   

   loanPurp  zipCode  pMIperct  mortInsType  avgCreditScore  fmacRate  \
0         1      724       0.0          0.0             733      4.32   
1         1      802       0.0          0.0             799      4.32   
2         2      109       0.0          0.0             753      4.32   
3         2      956       0.0          0.0             766      4.32   
4         1      760       0.0          0.0             707      4.32   

   fredRate  rateDiffPct  rank      deal  stateNumber  origYear  origMonth  \
0      2.67     0.099537     4   Premium            2      2014          2   
1      2.67     0.128472     4   Premium            5      2014          2   
2      2.67     0.099537     4   Premium           34      2014          2   
3      2.67    -0.016204     3  Discount            4      2014          2   
4      2.67     0.215278     4   Premium           43      2014          2   

   origWeek  
0         5  
1         5  
2         5  
3         5  
4         5

In [19]:
#df.to_csv('data/FannieAcq2000-19FMac30samp25MlReady13-10cut.csv')

# Pycaret - Setup with categorical definition

### Use feature engineering, polynomial features, trig feats, PCA

In [ ]:
# models=setup(df, target = 'rateDiffPct',
#              feature_interaction=True, feature_ratio=True,
#              polynomial_features=True, pca=True, pca_components=10,
#             numeric_features = ['origUPB','origLTV','numBorrowers','origDebtIncRatio',\
#                                'avgCreditScore'],
#             ignore_features = ['origIntRate','rank','deal'])

### Use feature engineering, polynomial features, trig feats

In [20]:
models=setup(df, target = 'rateDiffPct',
             feature_interaction=True, feature_ratio=True,
             polynomial_features=True,
            numeric_features = ['origUPB','origLTV','numBorrowers','origDebtIncRatio',\
                               'avgCreditScore'],
            ignore_features = ['origIntRate','rank','deal'])

 
Setup Succesfully Completed!


## Compare Models with blacklist exclusions

In [23]:
model_results=compare_models(blacklist = ['tr','ransac']) #,'lar','par','huber','llar','lasso','en','ridge','omp','br','svm'])
model_results

## Create models - store in variable

In [41]:
catboost =create_model('catboost', fold = 10)

MAE     MSE    RMSE      R2   RMSLE    MAPE
0     0.0444  0.0033  0.0572  0.5489  0.0477  0.6271
1     0.0440  0.0032  0.0563  0.5503  0.0463  0.4859
2     0.0425  0.0030  0.0547  0.5694  0.0456  0.4834
3     0.0442  0.0032  0.0565  0.5694  0.0471  0.5806
4     0.0425  0.0030  0.0548  0.5868  0.0452  0.4655
5     0.0448  0.0033  0.0572  0.5510  0.0474  0.4967
6     0.0442  0.0032  0.0565  0.5572  0.0463  0.5265
7     0.0442  0.0032  0.0565  0.5581  0.0470  0.4374
8     0.0437  0.0031  0.0560  0.5745  0.0464  0.5589
9     0.0436  0.0031  0.0560  0.5671  0.0460  0.4049
Mean  0.0438  0.0032  0.0562  0.5633  0.0465  0.5067
SD    0.0007  0.0001  0.0008  0.0117  0.0008  0.0640

In [40]:
catboost =create_model('catboost', fold = 50)

MAE     MSE    RMSE      R2   RMSLE    MAPE
0     0.0448  0.0034  0.0585  0.5319  0.0488  0.5551
1     0.0449  0.0032  0.0567  0.5472  0.0475  0.7093
2     0.0428  0.0031  0.0555  0.5808  0.0466  0.5539
3     0.0461  0.0036  0.0599  0.5127  0.0495  0.6825
4     0.0429  0.0030  0.0549  0.5785  0.0457  0.5920
5     0.0447  0.0033  0.0575  0.5483  0.0469  0.6331
6     0.0453  0.0034  0.0581  0.5427  0.0469  0.4772
7     0.0412  0.0028  0.0528  0.5737  0.0442  0.4422
8     0.0457  0.0033  0.0572  0.5201  0.0478  0.4227
9     0.0431  0.0031  0.0558  0.5676  0.0456  0.4770
10    0.0404  0.0027  0.0521  0.6057  0.0436  0.5748
11    0.0435  0.0032  0.0562  0.5515  0.0476  0.5279
12    0.0403  0.0026  0.0513  0.6152  0.0422  0.4623
13    0.0434  0.0031  0.0559  0.5472  0.0464  0.5812
14    0.0443  0.0033  0.0571  0.5318  0.0474  0.2525
15    0.0428  0.0030  0.0543  0.5946  0.0459  0.6056
16    0.0442  0.0033  0.0574  0.5500  0.0476  0.6724
17    0.0442  0.0030  0.0552  0.6052  0.0459  0.5462
18    0.0448  0.0033  0.0572  0.5451  0.0484  0.6690
19    0.0449  0.0033  0.0578  0.5559  0.0475  0.3565
20    0.0439  0.0032  0.0563  0.5771  0.0462  0.4803
21    0.0442  0.0033  0.0576  0.5620  0.0470  0.4298
22    0.0414  0.0028  0.0531  0.5939  0.0444  0.6728
23    0.0419  0.0028  0.0532  0.5980  0.0445  0.4098
24    0.0412  0.0029  0.0534  0.6011  0.0438  0.3757
25    0.0456  0.0034  0.0581  0.5491  0.0479  0.6522
26    0.0444  0.0032  0.0563  0.5444  0.0469  0.3748
27    0.0444  0.0032  0.0563  0.5905  0.0468  0.5628
28    0.0446  0.0032  0.0564  0.5482  0.0471  0.4580
29    0.0450  0.0034  0.0585  0.5207  0.0479  0.4507
30    0.0450  0.0034  0.0584  0.5050  0.0478  0.4591
31    0.0445  0.0031  0.0558  0.5747  0.0464  0.4940
32    0.0453  0.0033  0.0571  0.5704  0.0463  0.7151
33    0.0427  0.0031  0.0556  0.5821  0.0454  0.3721
34    0.0433  0.0031  0.0558  0.5446  0.0459  0.6187
35    0.0446  0.0032  0.0570  0.5315  0.0480  0.3881
36    0.0439  0.0032  0.0568  0.5582  0.0469  0.1701
37    0.0435  0.0031  0.0556  0.5559  0.0458  0.5141
38    0.0445  0.0031  0.0561  0.5669  0.0468  0.6487
39    0.0443  0.0032  0.0568  0.5805  0.0472  0.4648
40    0.0435  0.0031  0.0557  0.5644  0.0458  0.6074
41    0.0432  0.0030  0.0552  0.5882  0.0462  0.5135
42    0.0426  0.0030  0.0547  0.5856  0.0455  0.5822
43    0.0451  0.0033  0.0578  0.5576  0.0477  0.3877
44    0.0440  0.0032  0.0565  0.5768  0.0467  0.7229
45    0.0437  0.0032  0.0562  0.5770  0.0456  0.3717
46    0.0439  0.0032  0.0567  0.5343  0.0459  0.2008
47    0.0441  0.0032  0.0567  0.5652  0.0475  0.6691
48    0.0420  0.0029  0.0537  0.5837  0.0444  0.4467
49    0.0445  0.0032  0.0569  0.5719  0.0466  0.3528
Mean  0.0438  0.0032  0.0561  0.5633  0.0465  0.5072
SD    0.0013  0.0002  0.0017  0.0252  0.0014  0.1307

### Catboost fold results (not shown - reran in same cell)
* 10   0.5?
* 20   0.5633
* 50   0.5633

In [38]:
lightgbm = create_model('lightgbm', fold=50)

MAE     MSE    RMSE      R2   RMSLE    MAPE
0     0.0453  0.0035  0.0590  0.5242  0.0493  0.5556
1     0.0453  0.0033  0.0571  0.5417  0.0478  0.7061
2     0.0431  0.0031  0.0556  0.5794  0.0467  0.5243
3     0.0460  0.0036  0.0601  0.5104  0.0495  0.6603
4     0.0431  0.0030  0.0550  0.5775  0.0458  0.5854
5     0.0448  0.0033  0.0576  0.5464  0.0470  0.6326
6     0.0461  0.0035  0.0591  0.5266  0.0476  0.4634
7     0.0415  0.0029  0.0534  0.5639  0.0446  0.4331
8     0.0453  0.0033  0.0570  0.5232  0.0475  0.4218
9     0.0434  0.0032  0.0562  0.5612  0.0459  0.5077
10    0.0410  0.0027  0.0523  0.6021  0.0438  0.5667
11    0.0436  0.0032  0.0563  0.5495  0.0477  0.5205
12    0.0406  0.0026  0.0515  0.6130  0.0424  0.4798
13    0.0437  0.0032  0.0563  0.5409  0.0464  0.5506
14    0.0445  0.0033  0.0573  0.5293  0.0475  0.2506
15    0.0434  0.0030  0.0552  0.5823  0.0466  0.6202
16    0.0443  0.0033  0.0574  0.5492  0.0478  0.6738
17    0.0446  0.0031  0.0560  0.5944  0.0464  0.5278
18    0.0448  0.0033  0.0572  0.5449  0.0483  0.6580
19    0.0445  0.0033  0.0578  0.5561  0.0471  0.3318
20    0.0442  0.0032  0.0569  0.5681  0.0465  0.4579
21    0.0444  0.0034  0.0581  0.5536  0.0475  0.4415
22    0.0415  0.0028  0.0533  0.5904  0.0444  0.6420
23    0.0422  0.0029  0.0538  0.5899  0.0449  0.4094
24    0.0413  0.0029  0.0535  0.6000  0.0438  0.3654
25    0.0462  0.0035  0.0588  0.5388  0.0485  0.6491
26    0.0447  0.0032  0.0568  0.5370  0.0471  0.3824
27    0.0447  0.0032  0.0568  0.5831  0.0472  0.5680
28    0.0446  0.0032  0.0566  0.5456  0.0472  0.4116
29    0.0449  0.0035  0.0588  0.5161  0.0480  0.4444
30    0.0458  0.0035  0.0591  0.4933  0.0483  0.4770
31    0.0448  0.0032  0.0562  0.5676  0.0465  0.5051
32    0.0457  0.0033  0.0579  0.5591  0.0469  0.7350
33    0.0431  0.0031  0.0557  0.5798  0.0454  0.3488
34    0.0434  0.0031  0.0559  0.5435  0.0456  0.5851
35    0.0450  0.0033  0.0573  0.5261  0.0481  0.3743
36    0.0442  0.0032  0.0569  0.5559  0.0470  0.1644
37    0.0436  0.0031  0.0558  0.5517  0.0459  0.5092
38    0.0447  0.0032  0.0563  0.5646  0.0469  0.6321
39    0.0449  0.0033  0.0574  0.5720  0.0475  0.4595
40    0.0437  0.0031  0.0557  0.5635  0.0460  0.6084
41    0.0437  0.0031  0.0555  0.5840  0.0463  0.5283
42    0.0425  0.0030  0.0550  0.5810  0.0456  0.5669
43    0.0449  0.0033  0.0574  0.5632  0.0473  0.3925
44    0.0442  0.0032  0.0565  0.5757  0.0468  0.7314
45    0.0445  0.0033  0.0571  0.5632  0.0463  0.3735
46    0.0442  0.0032  0.0570  0.5303  0.0460  0.2137
47    0.0441  0.0032  0.0567  0.5667  0.0475  0.6532
48    0.0423  0.0029  0.0541  0.5776  0.0446  0.4351
49    0.0448  0.0033  0.0573  0.5668  0.0467  0.3367
Mean  0.0440  0.0032  0.0564  0.5585  0.0466  0.5014
SD    0.0013  0.0002  0.0017  0.0252  0.0014  0.1303

### LightGBM fold results (not shown - reran in same cell)
    * 10   0.5?
    * 20   0.5?
    * 50   0.5585

In [39]:
xgboost = create_model('xgboost', fold = 50)

MAE     MSE    RMSE      R2   RMSLE    MAPE
0     0.0463  0.0036  0.0602  0.5033  0.0501  0.5694
1     0.0473  0.0035  0.0590  0.5107  0.0489  0.7141
2     0.0443  0.0032  0.0569  0.5589  0.0477  0.6138
3     0.0475  0.0038  0.0615  0.4863  0.0505  0.7718
4     0.0450  0.0033  0.0570  0.5457  0.0473  0.5958
5     0.0467  0.0035  0.0591  0.5218  0.0480  0.6780
6     0.0469  0.0036  0.0602  0.5096  0.0483  0.5419
7     0.0423  0.0030  0.0545  0.5470  0.0451  0.4843
8     0.0471  0.0035  0.0588  0.4942  0.0487  0.4569
9     0.0448  0.0033  0.0578  0.5364  0.0466  0.5207
10    0.0424  0.0029  0.0542  0.5732  0.0451  0.6049
11    0.0447  0.0033  0.0571  0.5358  0.0481  0.5429
12    0.0424  0.0029  0.0535  0.5813  0.0440  0.4962
13    0.0448  0.0034  0.0581  0.5117  0.0477  0.5646
14    0.0457  0.0034  0.0587  0.5057  0.0482  0.2643
15    0.0451  0.0032  0.0569  0.5550  0.0476  0.6016
16    0.0457  0.0034  0.0587  0.5290  0.0486  0.7034
17    0.0461  0.0034  0.0583  0.5606  0.0477  0.5744
18    0.0463  0.0035  0.0590  0.5166  0.0496  0.7018
19    0.0460  0.0035  0.0591  0.5364  0.0481  0.3622
20    0.0454  0.0034  0.0580  0.5509  0.0472  0.4556
21    0.0460  0.0036  0.0598  0.5274  0.0481  0.4334
22    0.0426  0.0030  0.0547  0.5687  0.0450  0.6786
23    0.0439  0.0031  0.0555  0.5635  0.0463  0.4606
24    0.0433  0.0031  0.0554  0.5717  0.0451  0.4352
25    0.0469  0.0036  0.0597  0.5248  0.0493  0.6623
26    0.0458  0.0033  0.0578  0.5198  0.0476  0.3691
27    0.0460  0.0034  0.0584  0.5594  0.0481  0.5751
28    0.0458  0.0033  0.0574  0.5321  0.0475  0.4581
29    0.0463  0.0036  0.0601  0.4937  0.0486  0.4635
30    0.0466  0.0036  0.0602  0.4742  0.0490  0.4818
31    0.0465  0.0034  0.0581  0.5387  0.0478  0.4559
32    0.0473  0.0034  0.0586  0.5480  0.0470  0.7440
33    0.0442  0.0033  0.0571  0.5595  0.0462  0.3730
34    0.0450  0.0034  0.0579  0.5093  0.0472  0.6046
35    0.0455  0.0033  0.0578  0.5174  0.0484  0.3850
36    0.0460  0.0035  0.0588  0.5267  0.0483  0.2038
37    0.0455  0.0033  0.0575  0.5249  0.0472  0.5312
38    0.0461  0.0033  0.0577  0.5415  0.0480  0.6400
39    0.0464  0.0035  0.0587  0.5517  0.0484  0.4478
40    0.0454  0.0033  0.0574  0.5367  0.0471  0.6288
41    0.0455  0.0033  0.0575  0.5541  0.0475  0.5328
42    0.0445  0.0032  0.0569  0.5514  0.0467  0.5714
43    0.0464  0.0035  0.0593  0.5340  0.0484  0.3824
44    0.0457  0.0034  0.0585  0.5458  0.0484  0.7838
45    0.0452  0.0034  0.0582  0.5450  0.0469  0.3944
46    0.0452  0.0034  0.0580  0.5139  0.0465  0.2270
47    0.0461  0.0034  0.0587  0.5351  0.0486  0.6824
48    0.0440  0.0031  0.0560  0.5464  0.0462  0.4672
49    0.0462  0.0034  0.0584  0.5490  0.0472  0.3475
Mean  0.0454  0.0034  0.0579  0.5347  0.0476  0.5248
SD    0.0013  0.0002  0.0016  0.0235  0.0013  0.1350

## Tune Models

In [29]:
tuned_catboost = tune_model('catboost') # from fold=50

MAE     MSE    RMSE      R2   RMSLE    MAPE
0     0.0444  0.0033  0.0573  0.5472  0.0478  0.6147
1     0.0440  0.0032  0.0563  0.5500  0.0463  0.4859
2     0.0424  0.0030  0.0546  0.5704  0.0455  0.4754
3     0.0443  0.0032  0.0567  0.5665  0.0473  0.5763
4     0.0426  0.0030  0.0550  0.5829  0.0453  0.4552
5     0.0448  0.0033  0.0573  0.5491  0.0473  0.4890
6     0.0442  0.0032  0.0568  0.5535  0.0465  0.5395
7     0.0442  0.0032  0.0565  0.5588  0.0470  0.4425
8     0.0439  0.0032  0.0563  0.5706  0.0466  0.5480
9     0.0437  0.0032  0.0561  0.5655  0.0461  0.4086
Mean  0.0439  0.0032  0.0563  0.5614  0.0466  0.5035
SD    0.0007  0.0001  0.0008  0.0111  0.0008  0.0611

* Made no difference creating the model with fold= 10 vs. fold = 50 so use fold = 10

In [31]:
tuned_lightgbm = tune_model('lightgbm')

MAE     MSE    RMSE      R2   RMSLE    MAPE
0     0.0479  0.0037  0.0611  0.4858  0.0503  0.6673
1     0.0474  0.0036  0.0602  0.4861  0.0489  0.5574
2     0.0460  0.0035  0.0588  0.5020  0.0483  0.5115
3     0.0477  0.0037  0.0606  0.5037  0.0496  0.5962
4     0.0466  0.0035  0.0595  0.5128  0.0482  0.5162
5     0.0478  0.0037  0.0607  0.4943  0.0495  0.5219
6     0.0481  0.0037  0.0610  0.4842  0.0493  0.5313
7     0.0480  0.0037  0.0606  0.4911  0.0499  0.4410
8     0.0478  0.0037  0.0606  0.5020  0.0495  0.5754
9     0.0474  0.0037  0.0604  0.4964  0.0488  0.4457
Mean  0.0474  0.0036  0.0604  0.4958  0.0492  0.5364
SD    0.0006  0.0001  0.0007  0.0088  0.0006  0.0642

In [32]:
tuned_xgboost = tune_model('xgboost')

MAE     MSE    RMSE      R2   RMSLE    MAPE
0     0.0460  0.0036  0.0597  0.5085  0.0498  0.6511
1     0.0459  0.0035  0.0588  0.5097  0.0482  0.5120
2     0.0445  0.0033  0.0574  0.5258  0.0477  0.4845
3     0.0466  0.0036  0.0599  0.5152  0.0498  0.5903
4     0.0451  0.0034  0.0580  0.5357  0.0476  0.5077
5     0.0463  0.0035  0.0593  0.5166  0.0490  0.5207
6     0.0465  0.0036  0.0596  0.5074  0.0490  0.5561
7     0.0463  0.0035  0.0591  0.5169  0.0492  0.4543
8     0.0458  0.0035  0.0590  0.5287  0.0488  0.5824
9     0.0458  0.0035  0.0588  0.5235  0.0483  0.4413
Mean  0.0459  0.0035  0.0590  0.5188  0.0487  0.5301
SD    0.0006  0.0001  0.0007  0.0089  0.0007  0.0619

## Ensemble models

In [28]:
catboost_bagged = ensemble_model(catboost)

MAE     MSE    RMSE      R2   RMSLE    MAPE
0     0.0444  0.0033  0.0573  0.5474  0.0478  0.6229
1     0.0442  0.0032  0.0565  0.5477  0.0464  0.4915
2     0.0426  0.0030  0.0548  0.5679  0.0457  0.4841
3     0.0443  0.0032  0.0566  0.5677  0.0472  0.5791
4     0.0426  0.0030  0.0548  0.5856  0.0452  0.4571
5     0.0449  0.0033  0.0573  0.5486  0.0474  0.4938
6     0.0443  0.0032  0.0566  0.5553  0.0464  0.5284
7     0.0443  0.0032  0.0567  0.5555  0.0472  0.4329
8     0.0439  0.0032  0.0562  0.5710  0.0465  0.5604
9     0.0437  0.0032  0.0562  0.5649  0.0461  0.4175
Mean  0.0439  0.0032  0.0563  0.5612  0.0466  0.5068
SD    0.0007  0.0001  0.0008  0.0118  0.0008  0.0624

In [30]:
catboost_boosted = ensemble_model(catboost)

MAE     MSE    RMSE      R2   RMSLE    MAPE
0     0.0444  0.0033  0.0573  0.5474  0.0478  0.6229
1     0.0442  0.0032  0.0565  0.5477  0.0464  0.4915
2     0.0426  0.0030  0.0548  0.5679  0.0457  0.4841
3     0.0443  0.0032  0.0566  0.5677  0.0472  0.5791
4     0.0426  0.0030  0.0548  0.5856  0.0452  0.4571
5     0.0449  0.0033  0.0573  0.5486  0.0474  0.4938
6     0.0443  0.0032  0.0566  0.5553  0.0464  0.5284
7     0.0443  0.0032  0.0567  0.5555  0.0472  0.4329
8     0.0439  0.0032  0.0562  0.5710  0.0465  0.5604
9     0.0437  0.0032  0.0562  0.5649  0.0461  0.4175
Mean  0.0439  0.0032  0.0563  0.5612  0.0466  0.5068
SD    0.0007  0.0001  0.0008  0.0118  0.0008  0.0624

## Blend 3 chosen Models

#### First blend 'non-tuned' models

In [27]:
blend3 = blend_models(estimator_list = [catboost, lightgbm, xgboost])

MAE     MSE    RMSE      R2   RMSLE    MAPE
0     0.0447  0.0033  0.0575  0.5440  0.0479  0.6299
1     0.0444  0.0032  0.0567  0.5443  0.0464  0.4991
2     0.0428  0.0030  0.0550  0.5643  0.0457  0.4812
3     0.0446  0.0032  0.0569  0.5628  0.0473  0.5781
4     0.0429  0.0030  0.0551  0.5814  0.0453  0.4658
5     0.0450  0.0033  0.0574  0.5476  0.0473  0.4894
6     0.0447  0.0032  0.0570  0.5497  0.0465  0.5256
7     0.0446  0.0032  0.0568  0.5539  0.0471  0.4275
8     0.0442  0.0032  0.0564  0.5689  0.0466  0.5614
9     0.0440  0.0032  0.0564  0.5607  0.0462  0.4066
Mean  0.0442  0.0032  0.0565  0.5577  0.0466  0.5065
SD    0.0007  0.0001  0.0008  0.0115  0.0007  0.0653

#### Next blend tuned models

In [33]:
blendTuned3 = blend_models(estimator_list = [tuned_catboost, tuned_lightgbm, tuned_xgboost])

MAE     MSE    RMSE      R2   RMSLE    MAPE
0     0.0450  0.0034  0.0580  0.5368  0.0481  0.6316
1     0.0446  0.0032  0.0570  0.5393  0.0466  0.5032
2     0.0431  0.0031  0.0554  0.5579  0.0459  0.4753
3     0.0451  0.0033  0.0576  0.5517  0.0477  0.5729
4     0.0437  0.0031  0.0561  0.5668  0.0458  0.4769
5     0.0452  0.0033  0.0577  0.5423  0.0474  0.4965
6     0.0451  0.0033  0.0577  0.5383  0.0470  0.5267
7     0.0450  0.0033  0.0573  0.5455  0.0474  0.4320
8     0.0447  0.0033  0.0572  0.5570  0.0471  0.5537
9     0.0445  0.0032  0.0570  0.5526  0.0464  0.4176
Mean  0.0446  0.0033  0.0571  0.5488  0.0469  0.5086
SD    0.0007  0.0001  0.0008  0.0094  0.0007  0.0617

## Evaluate Models

In [ ]:
plot_model(lightgbm, plot = 'residuals')

In [ ]:
plot_model(lightgbm, plot = 'error')

In [ ]:
plot_model(lightgbm, plot='vc')

In [ ]:
plot_model(lightgbm, plot = 'feature')

In [ ]:
interpret_model(lightgbm, plot='correlation', feature='avgCreditScore')

In [ ]:
interpret_model(tuned_catboost)

In [ ]:
save_model()

In [ ]:
interpret_model(estimator=logreg_model, plot='reason')

In [ ]:
logreg_model2=create_model('xgboost')

In [ ]:
plot_model(estimator=logreg_model2, plot='feature')

In [ ]:
plot_model(estimator=logreg_model2)

In [ ]:
tunned_logreg_model=tune_model('catboost')

In [ ]:
interpret_model(estimator=tunned_logreg_model,plot='summary')

In [ ]:
plot_model(estimator=tunned_logreg_model,plot='feature')

In [ ]:
interpret_model(estimator=catboost, plot='correlation')

In [ ]:
interpret_model(estimator=catboost,plot='summary')